thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/translator/DetectionConfigMetricCache.java

```diff
@@ -0,0 +1,105 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.thirdeye.detection.yaml.translator;
+
+import java.text.ParseException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.collections4.MapUtils;
+import org.apache.pinot.thirdeye.common.time.TimeGranularity;
+import org.apache.pinot.thirdeye.datalayer.dto.DatasetConfigDTO;
+import org.apache.pinot.thirdeye.datalayer.dto.MetricConfigDTO;
+import org.apache.pinot.thirdeye.detection.DataProvider;
+import org.quartz.CronExpression;
+
+
+/**
+ * A data holder to store the processed information per metric
+ */
+class DetectionConfigMetricCache {
+
+  private static final String PROP_METRIC = "metric";
+  private static final String PROP_DATASET = "dataset";
+
+  private final Map<String, DetectionMetricProperties> metricCache = new HashMap<>();
+  private final DataProvider dataProvider;
+
+  DetectionConfigMetricCache(DataProvider provider) {
+    this.dataProvider = provider;
+  }
+
+  private void loadMetricCache(Map<String, Object> metricAlertConfigMap) {
+    String metricName = MapUtils.getString(metricAlertConfigMap, PROP_METRIC);
+    String datasetName = MapUtils.getString(metricAlertConfigMap, PROP_DATASET);
+
+    MetricConfigDTO metricConfig = this.dataProvider.fetchMetric(metricName, datasetName);
+    DatasetConfigDTO datasetConfig = this.dataProvider.fetchDatasets(Collections.singletonList(metricConfig.getDataset()))
+        .get(metricConfig.getDataset());
+    String cron = buildCron(datasetConfig.bucketTimeGranularity());
+
+    metricCache.put(metricName, new DetectionMetricProperties(cron, metricConfig, datasetConfig));
+  }
+
+  DatasetConfigDTO fetchDataset(Map<String, Object> metricAlertConfigMap) {
+    String metricName = MapUtils.getString(metricAlertConfigMap, PROP_METRIC);
+    if (!metricCache.containsKey(metricName)) {
+      loadMetricCache(metricAlertConfigMap);
+    }
+
+    return metricCache.get(metricName).getDatasetConfigDTO();
+  }
+
+  MetricConfigDTO fetchMetric(Map<String, Object> metricAlertConfigMap) {
+    String metricName = MapUtils.getString(metricAlertConfigMap, PROP_METRIC);
+    if (!metricCache.containsKey(metricName)) {
+      loadMetricCache(metricAlertConfigMap);
+    }
+
+    return metricCache.get(metricName).getMetricConfigDTO();
+  }
+
+  String fetchCron(Map<String, Object> metricAlertConfigMap) {
```

Comment:
What are the benefits of caching the cron expression?  Seems it is derived from the default granularity anyways. We can cache the metricConfig and datasetConfig and drive from that. Calling the `buildCron` method should be light.

We can also consider building the cache into the `DataProvider` so that it can be beneficial to more places.  Maybe leverage Guava's `LoadingCache`, it can handle cache expiration, eviction, etc.

Reply:
Although not implemented yet, the idea behind caching cron is to find the default cron in a composite alert. If we store a mapping of all the cron along with the metric/dataset, then we can potentially derive the default cron for the composite alert.

Reply:
It would be awesome to have a caching at the `DataProvider` level. We can discuss more about it but would like to keep it outside the scope of this PR.

Reply:
This is not a cache per se. It is mostly a data holder. The datastructure is created during translation and destroyed after that. I will rename this to avoid confusion.

Reply:
make sense.